### PR TITLE
Convert Stage to function component using hooks

### PR DESCRIPTION
**Pull request by 3mcd**
This PR is an example of how Stage might be refactored to a function component using the new hooks API introduced with React `0.0.17-alpha.0`. This will allow Stage to support function components as children that use hooks.

Below is an example of how the rotating bunny example might be rewritten using hooks:

```js
import React, { useState } from 'react';
import { usePixiTicker } from 'react-pixi-fiber';
import Bunny from "../Bunny";

export default function RotatingBunny(props) {
  const [rotation, setRotation] = useState(0);

  usePixiTicker(delta => {
    setRotation(prev => prev + delta * 0.1);
  });

  return <Bunny rotation={rotation} {...props} />;
}
```

**Todo**
- [x] Backwards compatibility
- [x] Duplicate unit tests for function variant of `Stage`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -95,6 +95,7 @@
     ],
     "transform": {
       "^.+\\.js$": "babel-jest"
-    }
+    },
+    "testURL": "http://localhost"
```
Comment (3mcd) [inline package.json]: Note: needed to add this for tests to pass (even before changes). See https://github.com/facebook/jest/issues/6766.

**Comment by michalochman** [comment]
Thanks @3mcd, this is rad! Thanks for approaching this. I'll have more in-depth look when I get a chance.

My first nitpick is that it completely relies on React 16.7 while there certainly are users that will not be able to upgrade to that version. This library already supports multiple React versions at the same time:
- pre 16.3 (Legacy Context API)
- 16.3+ (New Context API)
- 16.5+ (change of `ReactWorkTags` in `react-reconciler`/`react-dom`)

so I would expect hooks and any other new features to be implemented with backward-compatibility in mind, unless of course there are so many changes to React core that it's hard to maintain (which is not the case at the moment). That's how Legacy / New Context API is handled currently, have a look at [AppProvider implementation](https://github.com/michalochman/react-pixi-fiber/blob/41c341d8c8a851e9cf3678258b075ac9cb47bf17/src/AppProvider.js).

File: src/Stage.js [hunk 2]
```diff
@@ -55,60 +56,72 @@ export const includingCanvasProps = key => !includingDisplayObjectProps(key) &&
 export const getCanvasProps = props => filterByKey(props, includingCanvasProps);
 export const getDisplayObjectProps = props => filterByKey(props, includingDisplayObjectProps);
 
-class Stage extends React.Component {
-  componentDidMount() {
-    const { children, height, options, width } = this.props;
+export let appTestHook = null;
```
Comment (3mcd) [inline src/Stage.js]: We can no longer rely on the instance property `_app` on Stage for tests now that it has been refactored into a function component. This is a temporary solution to expose the most recently created pixi application instance so tests can hook into Stage internals.

**Comment by 3mcd** [comment]
@michalochman Awesome! Thanks for the feedback and direction. Looking forward to a more in depth review. I must admit I'm still learning the new hooks API so I'm sure we can make some improvements to the implementation. I'll start working on the backwards compatibility changes you mentioned!

File: src/Stage.js [hunk 3]
```diff
@@ -55,62 +56,133 @@ export const includingCanvasProps = key => !includingDisplayObjectProps(key) &&
...
 
-    this._app = new PIXI.Application({
+export function usePixiApp(props) {
```
Comment (michalochman) [inline src/Stage.js]: This could be moved to `hooks.js` file as well.

**Comment by nstolpe** [comment]
@3mcd @michalochman I've made a fork of this PR and resolved most of the conflicts. I haven't tested the build yet, and some tests still need updating, but if it will help adding hook support, I can open a PR on this branch or master and then resolve whatever's still outstanding.

**Comment by michalochman** [comment]
@nstolpe is this it: https://github.com/nstolpe/react-pixi-fiber/pull/2?

**Comment by nstolpe** [comment]
@michalochman yeah, that's it. I'm going to do some testing, it's possible it won't work at all. But maybe it will :man_shrugging:

**Comment by michalochman** [comment]
@3mcd @nstolpe FYI I have updated react dependencies in react-pixi-fiber@0.9.0, that should fix some issues that were reported around usage with hooks

**Comment by nstolpe** [comment]
@michalochman cool, thanks! I walked through a bunch of the code for a few days before I tried updating react-reconciler, then that got everything working for me even without any other changes. At least I got a little familiar with the internals.

**Comment by michalochman** [comment]
@nstolpe would you be able to open new PR then? This one has many conflicts.

**Comment by nstolpe** [comment]
@michalochman sure. I've been testing off master, and basic hook functionality is working fine...at least my apps will load without failing because hooks can't be used. I am seeing an issue with `useEffect` though. It's the same error in [this thread](https://github.com/facebook/react/issues/15647), and I wonder if it has something to do w/ custom renderers since I've had the same problem as in [this comment](https://github.com/facebook/react/issues/15647#issuecomment-493506786) which references `react-konva`.

I'll do some testing with my branch that has @3mcd's changes re-incorporated and see if maybe a function component stage will resolve it.

If everything is working fine (or at least as fine as it can be if that useEffect issue is external) with master and hooks, do you want to leave stage as is and just incorporate the hooks from the original PR? Or still incorporate everything?

**Comment by nstolpe** [comment]
@michalochman I've [opened  PR](https://github.com/michalochman/react-pixi-fiber/pull/127). Tests still need updating/adding, and I still need to do the testing mentioned in my previous comment, but let me know if you have any thoughts.

**Comment by michalochman** [comment]
@nstolpe do you still have this `useEffect` issue after downgrading `scheduler` to 0.13.6?

**Comment by nstolpe** [comment]
@michalochman I'd tried downgrading to 0.13.6 before, but I must not have been thorough enough. I just tested off master again and the effect hook is fine (no callback is not a function error).

**Comment by 3mcd** [comment]
Closing this PR as hooks are now working as expected.

**Pull request closed without merging**